File: src/main/java/org/tindalos/principle/domain/detector/submoduleboundaries/SubmodulesFactory.java
package org.tindalos.principle.domain.detector.submoduleboundaries;

import java.util.List;

import org.tindalos.principle.domain.core.Package;
import org.tindalos.principle.domain.detector.adp.PackageStructureBuilder;
import org.tindalos.principle.domain.expectations.SubmodulesDefinitionLocation;

public class SubmodulesFactory {
    
    private final PackageStructureBuilder packageStructureBuilder;
    private final SubmoduleDefinitionsProvider submoduleDefinitionsProvider;
    private final SubmoduleFactory submoduleFactory;
    

    public SubmodulesFactory(PackageStructureBuilder packageStructureBuilder,
            SubmoduleDefinitionsProvider submoduleDefinitionsProvider, SubmoduleFactory submoduleFactory) {
        this.packageStructureBuilder = packageStructureBuilder;
        this.submoduleDefinitionsProvider = submoduleDefinitionsProvider;
        this.submoduleFactory = submoduleFactory;
    }

    public List<Submodule> buildSubmodules(SubmodulesDefinitionLocation submodulesDefinitionLocation,
            List<Package> packages, String basePackageName) {
       
        SubmoduleDefinitions submoduleDefinitions = submoduleDefinitionsProvider.readSubmoduleDefinitions(submodulesDefinitionLocation);
        Package basePackage = packageStructureBuilder.build(packages, basePackageName);

        return submoduleFactory.buildModules(submoduleDefinitions, basePackage.toMap());
    }

}

File: src/main/java/org/tindalos/principle/domain/detector/submodulesblueprint/SubmodulesFactory.scala
package org.tindalos.principle.domain.detector.submodulesblueprint

import org.tindalos.principle.domain.core.{PackageReference, Package}

class SubmodulesFactory(
  private val packageStructureBuilder: (List[Package], String) => Package,
  private val submoduleDefinitionsProvider: SubmoduleDefinitionsProvider,
  submoduleFactory: (SubmoduleDefinitions, Map[PackageReference, Package]) => Set[Submodule]) {

  def buildSubmodules(submodulesDefinitionLocation: String, packages: List[Package], basePackageName: String) = {

    val submoduleDefinitions = submoduleDefinitionsProvider.readSubmoduleDefinitions(submodulesDefinitionLocation, basePackageName)
    val basePackage = packageStructureBuilder(packages, basePackageName)

    submoduleFactory(submoduleDefinitions, basePackage.toMap())
  }
}